File: URL-Shortner/fastapi-service/app/routers/redirect.py
from datetime import datetime, timezone

from fastapi import APIRouter, BackgroundTasks
from fastapi.responses import JSONResponse, RedirectResponse

from app import analytics, cache
from app.config import settings
from app.db import get_pool

router = APIRouter()


def _is_expired(expires_at: datetime | None) -> bool:
    return expires_at is not None and expires_at < datetime.now(timezone.utc)


async def _record_click_sync_bg(short_code: str) -> None:
    # Runs after the redirect response has already been sent to the client
    # (Starlette flushes the response before executing background tasks) -
    # this is what keeps analytics off the redirect's critical path.
    pool = get_pool()
    async with pool.acquire() as conn:
        async with conn.transaction():
            await conn.execute(
                "UPDATE links SET click_count = click_count + 1 WHERE short_code = $1",
                short_code,
            )
            await conn.execute(
                "INSERT INTO click_events (short_code) VALUES ($1)", short_code
            )


def _schedule_click(background_tasks: BackgroundTasks, short_code: str) -> None:
    if settings.analytics_mode == "buffered":
        analytics.record_click(short_code)  # pure in-memory, no I/O at all
    else:
        background_tasks.add_task(_record_click_sync_bg, short_code)


@router.get("/{code}")
async def redirect_to_long_url(code: str, background_tasks: BackgroundTasks):
    cached = await cache.get_link(code)
    if cached is not None:
        if _is_expired(cached["expires_at"]):
            return JSONResponse(status_code=410, content={"error": "expired"})
        _schedule_click(background_tasks, code)
        return RedirectResponse(url=cached["long_url"], status_code=302)

    # Cache miss: fall back to Postgres.
    pool = get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT long_url, expires_at FROM links WHERE short_code = $1", code
        )

    if row is None:
        return JSONResponse(status_code=404, content={"error": "not_found"})

    if _is_expired(row["expires_at"]):
        # Don't cache dead entries - nothing gained from remembering a 410.
        return JSONResponse(status_code=410, content={"error": "expired"})

    await cache.set_link(code, row["long_url"], row["expires_at"])
    _schedule_click(background_tasks, code)
    return RedirectResponse(url=row["long_url"], status_code=302)

File: URL-Shortner/go-service/internal/handlers/redirect.go
package handlers

import (
	"context"
	"errors"
	"log"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"urlshortener/internal/cache"
)

type RedirectHandler struct {
	Pool  *pgxpool.Pool
	Cache *cache.Client
}

func isExpired(expiresAt *time.Time) bool {
	return expiresAt != nil && expiresAt.Before(time.Now().UTC())
}

// recordClick runs in its own goroutine after the redirect response has
// already been written - this is what keeps analytics off the redirect's
// critical path (mirrors FastAPI's BackgroundTasks on the other stack).
func (h *RedirectHandler) recordClick(shortCode string) {
	ctx := context.Background()
	tx, err := h.Pool.Begin(ctx)
	if err != nil {
		log.Printf("record click: begin: %v", err)
		return
	}
	defer tx.Rollback(ctx)

	if _, err := tx.Exec(ctx, `UPDATE links SET click_count = click_count + 1 WHERE short_code = $1`, shortCode); err != nil {
		log.Printf("record click: update: %v", err)
		return
	}
	if _, err := tx.Exec(ctx, `INSERT INTO click_events (short_code) VALUES ($1)`, shortCode); err != nil {
		log.Printf("record click: insert: %v", err)
		return
	}
	if err := tx.Commit(ctx); err != nil {
		log.Printf("record click: commit: %v", err)
	}
}

func (h *RedirectHandler) Redirect(w http.ResponseWriter, r *http.Request) {
	code := chi.URLParam(r, "code")

	cached, err := h.Cache.GetLink(r.Context(), code)
	if err == nil && cached != nil {
		if isExpired(cached.ExpiresAt) {
			writeError(w, http.StatusGone, "expired")
			return
		}
		http.Redirect(w, r, cached.LongURL, http.StatusFound)
		go h.recordClick(code)
		return
	}

	// Cache miss (or cache error - fail open to Postgres): fall back to the DB.
	var longURL string
	var expiresAt *time.Time
	dbErr := h.Pool.QueryRow(r.Context(), `SELECT long_url, expires_at FROM links WHERE short_code = $1`, code).
		Scan(&longURL, &expiresAt)
	if dbErr != nil {
		if errors.Is(dbErr, pgx.ErrNoRows) {
			writeError(w, http.StatusNotFound, "not_found")
			return
		}
		writeError(w, http.StatusInternalServerError, "internal_error")
		return
	}

	if isExpired(expiresAt) {
		// Don't cache dead entries - nothing gained from remembering a 410.
		writeError(w, http.StatusGone, "expired")
		return
	}

	if err := h.Cache.SetLink(r.Context(), code, longURL, expiresAt); err != nil {
		log.Printf("cache set: %v", err)
	}

	http.Redirect(w, r, longURL, http.StatusFound)
	go h.recordClick(code)
}
